### DISCUSSION TITLE
[Solution] Was able to get Plugins Installed/Working (Thanks 'Alu fers' on Telegram)

### DISCUSSION THREAD

**Discussion by ChristilawD** [feelfreelinux/octo4a#247] [General]
So, I installed this on my old Samsing Galaxy Note 9. Works like a champ, but I was not able to install several plugins I wanted, specifically Arc Welder.

Thanks for the assist via Telegram Alu fers! 👯 

Anyhow, the errors I was getting were Build Failures because GCC (as well as several other tools) were not installed.

You can fix this pretty easy.

SSH into the Print Server:
1. Enable the SSH Server on the Octo4a App
2. Create a password
3. Note the Port (default 8022)
4. From your favorite command shell, you need to be able to SSH
5. Linux: ssh -p 8022 octoprint@xxx.xxx.xxx.xxx <--- IP Address as your Print Server (look on the phone)
6. Password: Will be the one you set in the app

Now, you are on the server and should see a prompt:

localhost:~$

From here, you need to install the missing tools:

- **Install Dev Tools**: apk add build-base
- **Install Python3-Dev**: apk add python3-dev

If you get some errors, here are some other packages that were installed (that should be included in the above) for troubleshooting:

- **Wheel**: pip install wheel
- **GCC**: apk add gcc-arm-none-eabi
- **BIN Utils**: apk add binutils-arm-none-eabi

**Comment by alufers** [comment]
:)

**Comment by aarons0n** [comment]
Thanks for the good description. I follow it and tried to install the smart filament sensor plugin, but unfortunately just got an error message.
Any idea what I can do?

> pip install "https://github.com/Royrdan/Octoprint-Smart-Filament-Sensor/archive/master.zip"
Collecting https://github.com/Royrdan/Octoprint-Smart-Filament-Sensor/archive/master.zip
ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-tg01fp7i/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-tg01fp7i/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-cv4_kcoh
       cwd: /tmp/pip-req-build-tg01fp7i/
  Complete output (53 lines):
  Found packages: {'octoprint_smart_filament_sensor', 'octoprint_smart_filament_sensor.data'}
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib
  creating build/lib/octoprint_smart_filament_sensor
  copying octoprint_smart_filament_sensor/__init__.py -> build/lib/octoprint_smart_filament_sensor
  copying octoprint_smart_filament_sensor/filament_motion_sensor_connection_check.py -> build/lib/octoprint_smart_filament_sensor
  copying octoprint_smart_filament_sensor/filament_motion_sensor_timeout_detection.py -> build/lib/octoprint_smart_filament_sensor
  creating build/lib/octoprint_smart_filament_sensor/data
  copying octoprint_smart_filament_sensor/data/__init__.py -> build/lib/octoprint_smart_filament_sensor/data
  running egg_info
  creating Octoprint_Smart_Filament_Sensor.egg-info
  writing Octoprint_Smart_Filament_Sensor.egg-info/PKG-INFO
  writing dependency_links to Octoprint_Smart_Filament_Sensor.egg-info/dependency_links.txt
  writing entry points to Octoprint_Smart_Filament_Sensor.egg-info/entry_points.txt
  writing requirements to Octoprint_Smart_Filament_Sensor.egg-info/requires.txt
  writing top-level names to Octoprint_Smart_Filament_Sensor.egg-info/top_level.txt
  writing manifest file 'Octoprint_Smart_Filament_Sensor.egg-info/SOURCES.txt'
  reading manifest file 'Octoprint_Smart_Filament_Sensor.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  writing manifest file 'Octoprint_Smart_Filament_Sensor.egg-info/SOURCES.txt'
  creating build/lib/octoprint_smart_filament_sensor/static
  creating build/lib/octoprint_smart_filament_sensor/static/js
  copying octoprint_smart_filament_sensor/static/js/smartfilamentsensor_sidebar.js -> build/lib/octoprint_smart_filament_sensor/static/js
  copying octoprint_smart_filament_sensor/static/js/smartfilamentsensor_settings.js -> build/lib/octoprint_smart_filament_sensor/static/js
  creating build/lib/octoprint_smart_filament_sensor/templates
  copying octoprint_smart_filament_sensor/templates/smartfilamentsensor_sidebar.jinja2 -> build/lib/octoprint_smart_filament_sensor/templates
  copying octoprint_smart_filament_sensor/templates/smartfilamentsensor_sidebar_header.jinja2 -> build/lib/octoprint_smart_filament_sensor/templates
  copying octoprint_smart_filament_sensor/templates/smartfilamentsensor_settings.jinja2 -> build/lib/octoprint_smart_filament_sensor/templates
  installing to build/bdist.linux-aarch64/wheel
  running install
  running install_lib
  creating build/bdist.linux-aarch64
  creating build/bdist.linux-aarch64/wheel
  creating build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor
  copying build/lib/octoprint_smart_filament_sensor/filament_motion_sensor_connection_check.py -> build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor
  copying build/lib/octoprint_smart_filament_sensor/__init__.py -> build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor
  copying build/lib/octoprint_smart_filament_sensor/filament_motion_sensor_timeout_detection.py -> build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor
  creating build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor/templates
  copying build/lib/octoprint_smart_filament_sensor/templates/smartfilamentsensor_sidebar.jinja2 -> build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor/templates
  copying build/lib/octoprint_smart_filament_sensor/templates/smartfilamentsensor_settings.jinja2 -> build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor/templates
  copying build/lib/octoprint_smart_filament_sensor/templates/smartfilamentsensor_sidebar_header.jinja2 -> build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor/templates
  creating build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor/static
  creating build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor/static/js
  copying build/lib/octoprint_smart_filament_sensor/static/js/smartfilamentsensor_settings.js -> build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor/static/js
  copying build/lib/octoprint_smart_filament_sensor/static/js/smartfilamentsensor_sidebar.js -> build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor/static/js
  creating build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor/data
  copying build/lib/octoprint_smart_filament_sensor/data/__init__.py -> build/bdist.linux-aarch64/wheel/octoprint_smart_filament_sensor/data
  running install_egg_info
  Copying Octoprint_Smart_Filament_Sensor.egg-info to build/bdist.linux-aarch64/wheel/Octoprint_Smart_Filament_Sensor-1.2-py3.9.egg-info
  error: [Errno 13] Permission denied: 'build/bdist.linux-aarch64/wheel/Octoprint_Smart_Filament_Sensor-1.2-py3.9.egg-info/entry_points.txt'
  ----------------------------------------
  ERROR: Failed building wheel for Octoprint-Smart-Filament-Sensor
  Running setup.py clean for Octoprint-Smart-Filament-Sensor
Failed to build Octoprint-Smart-Filament-Sensor
Installing collected packages: Octoprint-Smart-Filament-Sensor
    Running setup.py install for Octoprint-Smart-Filament-Sensor ... error
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-tg01fp7i/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-tg01fp7i/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-5borlpir/install-record.txt --single-version-externally-managed --compile --install-headers /usr/include/python3.9/Octoprint-Smart-Filament-Sensor
         cwd: /tmp/pip-req-build-tg01fp7i/
    Complete output (53 lines):
    Found packages: {'octoprint_smart_filament_sensor.data', 'octoprint_smart_filament_sensor'}
    running install
    running build
    running build_py
    creating build
    creating build/lib
    creating build/lib/octoprint_smart_filament_sensor
    creating build/lib/octoprint_smart_filament_sensor/data
    copying octoprint_smart_filament_sensor/data/__init__.py -> build/lib/octoprint_smart_filament_sensor/data
    copying octoprint_smart_filament_sensor/__init__.py -> build/lib/octoprint_smart_filament_sensor
    copying octoprint_smart_filament_sensor/filament_motion_sensor_connection_check.py -> build/lib/octoprint_smart_filament_sensor
    copying octoprint_smart_filament_sensor/filament_motion_sensor_timeout_detection.py -> build/lib/octoprint_smart_filament_sensor
    running egg_info
    creating Octoprint_Smart_Filament_Sensor.egg-info
    writing Octoprint_Smart_Filament_Sensor.egg-info/PKG-INFO
    writing dependency_links to Octoprint_Smart_Filament_Sensor.egg-info/dependency_links.txt
    writing entry points to Octoprint_Smart_Filament_Sensor.egg-info/entry_points.txt
    writing requirements to Octoprint_Smart_Filament_Sensor.egg-info/requires.txt
    writing top-level names to Octoprint_Smart_Filament_Sensor.egg-info/top_level.txt
    writing manifest file 'Octoprint_Smart_Filament_Sensor.egg-info/SOURCES.txt'
    reading manifest file 'Octoprint_Smart_Filament_Sensor.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    writing manifest file 'Octoprint_Smart_Filament_Sensor.egg-info/SOURCES.txt'
    creating build/lib/octoprint_smart_filament_sensor/static
    creating build/lib/octoprint_smart_filament_sensor/static/js
    copying octoprint_smart_filament_sensor/static/js/smartfilamentsensor_sidebar.js -> build/lib/octoprint_smart_filament_sensor/static/js
    copying octoprint_smart_filament_sensor/static/js/smartfilamentsensor_settings.js -> build/lib/octoprint_smart_filament_sensor/static/js
    creating build/lib/octoprint_smart_filament_sensor/templates
    copying octoprint_smart_filament_sensor/templates/smartfilamentsensor_sidebar.jinja2 -> build/lib/octoprint_smart_filament_sensor/templates
    copying octoprint_smart_filament_sensor/templates/smartfilamentsensor_sidebar_header.jinja2 -> build/lib/octoprint_smart_filament_sensor/templates
    copying octoprint_smart_filament_sensor/templates/smartfilamentsensor_settings.jinja2 -> build/lib/octoprint_smart_filament_sensor/templates
    running install_lib
    creating /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor
    creating /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/templates
    copying build/lib/octoprint_smart_filament_sensor/templates/smartfilamentsensor_sidebar.jinja2 -> /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/templates
    copying build/lib/octoprint_smart_filament_sensor/templates/smartfilamentsensor_settings.jinja2 -> /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/templates
    copying build/lib/octoprint_smart_filament_sensor/templates/smartfilamentsensor_sidebar_header.jinja2 -> /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/templates
    creating /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/static
    creating /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/static/js
    copying build/lib/octoprint_smart_filament_sensor/static/js/smartfilamentsensor_settings.js -> /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/static/js
    copying build/lib/octoprint_smart_filament_sensor/static/js/smartfilamentsensor_sidebar.js -> /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/static/js
    copying build/lib/octoprint_smart_filament_sensor/__init__.py -> /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor
    creating /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/data
    copying build/lib/octoprint_smart_filament_sensor/data/__init__.py -> /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/data
    copying build/lib/octoprint_smart_filament_sensor/filament_motion_sensor_timeout_detection.py -> /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor
    copying build/lib/octoprint_smart_filament_sensor/filament_motion_sensor_connection_check.py -> /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor
    byte-compiling /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/__init__.py to __init__.cpython-39.pyc
    byte-compiling /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/data/__init__.py to __init__.cpython-39.pyc
    byte-compiling /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/filament_motion_sensor_timeout_detection.py to filament_motion_sensor_timeout_detection.cpython-39.pyc
    byte-compiling /usr/lib/python3.9/site-packages/octoprint_smart_filament_sensor/filament_motion_sensor_connection_check.py to filament_motion_sensor_connection_check.cpython-39.pyc
    running install_egg_info
    Copying Octoprint_Smart_Filament_Sensor.egg-info to /usr/lib/python3.9/site-packages/Octoprint_Smart_Filament_Sensor-1.2-py3.9.egg-info
    error: [Errno 13] Permission denied: '/usr/lib/python3.9/site-packages/Octoprint_Smart_Filament_Sensor-1.2-py3.9.egg-info/SOURCES.txt'
    ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-tg01fp7i/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-tg01fp7i/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-5borlpir/install-record.txt --single-version-externally-managed --compile --install-headers /usr/include/python3.9/Octoprint-Smart-Filament-Sensor Check the logs for full command output.

**Reply by ChristilawD** [reply]
Yeah, this one is not going to be as easy as the others. I would suggest you contact the Plugin Developer.

Ultimately, the issue you are having is Permissions errors on some files:

`error: [Errno 13] Permission denied: 'build/bdist.linux-aarch64/wheel/Octoprint_Smart_Filament_Sensor-1.2-py3.9.egg-info/entry_points.txt'`
and
`error: [Errno 13] Permission denied: '/usr/lib/python3.9/site-packages/Octoprint_Smart_Filament_Sensor-1.2-py3.9.egg-info/SOURCES.txt'`

This normally happens when the permissions on the directory the script is being executed from are incorrect.

Now, if your Linux-Foo is strong, you can always drop into an SSH session on the print server (described above) and try and manually install the plugin (if they let you that is, because I have not really messed with this that much, future project). Just note that this is running on Alpine Linux. For normal Linux Users, this isn't a "normal" distro as it's primary focus is to be small, light weight and used primarily in Container Implementations.

Here's a good starting place to learn Alpine commands: https://www.cyberciti.biz/faq/10-alpine-linux-apk-command-examples/

Sorry I could not offer much more, but I will see if I can dig into this a bit more this weekend when I have some more time.
